mysql/mysqldump.go

```diff
@@ -0,0 +1,578 @@
+	// There are no methods in pincap parser to retirieve ID and mods
+	// We will process columnType eg:'varchar(40)' and split ID from the string.
+	// We retrieve mods using regex expression and convert it to INT64
+	id := columnType
+	var mods []int64
+	if strings.Contains(columnType, "(") {
+		id = strings.Split(columnType, "(")[0]
+		// For 'set' and 'enum' datatypes, values provided are not maxLength
+		if id == "set" || id == "enum" {
+			return id, nil
+		}
+		values := valuesRegexp.FindString(columnType)
+		strMods := strings.Split(values[1:len(values)-1], ",")
+		for _, i := range strMods {
+			j, err := strconv.ParseInt(i, 10, 64)
+			if err != nil {
+				panic(err)
+			}
+			mods = append(mods, j)
+		}
+	}
+	// 'BINARY' keyword will be added to all blob datatypes by parser.
+	// Eg: mediumblob BINARY. It needs to be ignored to retrieve ID
+	if strings.Contains(id, " ") {
+		id = strings.Split(columnType, " ")[0]
+	}
+	return id, mods
+}
+
+// handleParseError handles error while parsing mysqldump
+// statements and attempts at creating parsable query.
+// Error can be due to insert statement or unsupported Spatial
+// datatypes in create statement.
+func handleParseError(conv *internal.Conv, query string, err error, l [][]byte) ([]ast.StmtNode, bool) {
+	// Check if error is due to Insert statement.
+	insertStmt := insertRegexp.FindString(query)
+	if insertStmt != "" {
+		// Likely causes of failing to parse Insert statement:
+		// a) Due to some invalid value
+		// b) Query size is more than what pingcap parser could handle (more than 40MB in size)
+		return handleInsertStatement(conv, query, insertStmt)
+	}
+
+	// Check if error is due to spatial datatype as it is not supported by Pingcap parser.
+	var isSpatial bool
+	for _, spatial := range MysqlSpatialDataTypes {
+		if strings.Contains(strings.ToLower(err.Error()), spatial) && isSpatial == false {
+			isSpatial = true
+			if conv.SchemaMode() {
+				conv.Unexpected(fmt.Sprintf("Unsupported datatype '%s' encountered while parsing following statement at line number %d : \n%s", spatial, len(l), query))
+				internal.VerbosePrintf("Converting datatype '%s' to 'Text' and retrying to parse the statement\n", spatial)
+			}
+		}
+	}
+	if isSpatial {
+		return handleSpatialDatatype(conv, query, l)
+	}
+
+	return nil, true
+}
+
+// handleInsertStatement handles error in parsing the insert statement.
+// We deal with this case by extracting all rows and creating
+// extended insert statements. Then we parse one Insert statement
+// at a time, ensuring no size issue and skipping only invalid entries.
+func handleInsertStatement(conv *internal.Conv, query, insertStmt string) ([]ast.StmtNode, bool) {
+	var stmts []ast.StmtNode
+	values := valuesRegexp.FindAllString(query, -1)
+	if len(values) == 0 {
+		return nil, true
+	}
+	for _, value := range values {
+		query = insertStmt + value + ";"
+		newTree, _, err := parser.New().Parse(query, "", "")
+		if err != nil {
+			if conv.SchemaMode() {
```

Comment:
Why the conditional on SchemaMode?

Should this be conditional on DataMode (since we only care about the processing of data inserts during data mode)?

Separately, how about skipping all of this processing unless we're in DataMode?

Reply:
Conditional SchemaMode because as mentioned on earlier comment, we are getting row-stats while processing insert statements in schema mode. 
Condition is to avoid double counting and it can also be changed to conditional DataMode(). I have added unexpected condition in SchemaMode() so that if `verbose` is set then user can get unexpected condition before going to DataMode()

Reply:
Ah yes, of course.
Sorry, I'd forgotten that we need to count rows during schema mode so we can provide a progress percentage when in data mode.